src/west/_bootstrap/main.py

```diff
@@ -195,6 +196,14 @@ def init_bootstrap(directory, args):
     clone(args.manifest_url, args.manifest_rev,
           os.path.join(directory, WEST_DIR, MANIFEST))
 
+    # Create an initial configuration file
+
+    config = configparser.ConfigParser()
+    config['west'] = {'revision': args.west_rev}
+    config['manifest'] = {'revision': args.manifest_rev}
+    with open(os.path.join(directory, WEST_DIR, 'config'), 'w') as f:
```

Comment:
Isn't that how Git behaves? It's the only thing that makes sense to me.

What should it do instead? Ignore the command-line parameters if there are system-wide settings?

Reply:
Git overwrites the other configs. I tried this by adding:

```
[core]
filemode = false
```
in my `~/.gitconfig`. After cloning a new repo, I get
```
filemode = true
```

in `repo/.git/config`

Reply:
Then again this might not be the best [example](https://stackoverflow.com/a/30394903/337201)

Reply:
Yeah, repository settings override system-wide settings in Git. The system-wide settings are more like defaults.

It'd get messy if they didn't, because there'd be no way to override a setting that happens to be in a system-wide configuration file.

Reply:
What you could do would be to skip adding those entries to `west/conf` if no revisions are passed to `west init`. Then system-wide defaults could kick in.

Comes down to what you think the user wants when they run `west init` without specifying the revision. Is it the defaults for the parameters, or is the system-wide defaults?

IMO, it's much less confusing it means the defaults for those parameters (otherwise, we'd have to change the documentation too). The only real "save" if it meant some system-wide default revision would be that people wouldn't have to give the revision they want explicitly, and it just doesn't seem worth it.

Reply:
@mbolivar so your expectation is that `west init` follows the url and revisions present in your global or system configuration file in the case where the user has not specified one in the command line?

Reply:
> @mbolivar so your expectation is that `west init` follows the url and revisions present in your global or system configuration file in the case where the user has not specified one in the command line?

That's my expectation, yes. I think if the user has explicitly bothered to set up a config file in a non-default location and tried to configure west or the manifest there, they clearly expect that to have some effect.

> IMO, it's much less confusing it means the defaults for those parameters (otherwise, we'd have to change the documentation too). The only real "save" if it meant some system-wide default revision would be that people wouldn't have to give the revision they want explicitly if it's different from master, and it just doesn't seem worth it.

I actually think it is more confusing to do what you have described, given that you've decided to put these settings in the west configuration file in this implementation. The config file behavior should be consistent, rather than splitting some config into "it's okay for these to be global" and other config into "this is local only".

If the west and manifest URLs don't fit this pattern, then I don't think they should come from the West config file system.

Reply:
>If the west and manifest URLs don't fit this pattern, then I don't think they should come from the West config file system.

I agree with that, and also I am puzzled by the fact that only the revision is there and not the URL.
But @mbolivar, what would you do in the case of command-line options present? I would personally assume that command-line options take the highest priority in this case, since the user is specifically calling west with a set of URLs and revisions.

Reply:
> But @mbolivar, what would you do in the case of command-line options present? I would personally assume that command-line options take the highest priority in this case, since the user is specifically calling west with a set of URLs and revisions.

I agree. In increasing order of precedence, I think the west and manifest URL and revision should be determined by:

1. Defaults hard-coded into the bootstrapper
2. Site-wide config files (if any)
3. User-wide config files
4. Existing wworkdir west config (this only applies in the "reinit" case, i.e. if ``west init`` is invoked in an existing wworkdir)
5. Explicit ``west init`` command line options

Reply:
>Comes down to what you think the user wants when they run west init without specifying the revision. Is it the defaults for the parameters, or the system-wide defaults?

I think the user wants the system-wide defaults there, otherwise he wouldn't have bothered to fill those in in the config file. Why not print it?

```
west init .
Cloning west from URL:<url> at revision:<revision>...
Cloning manifest from URL:<url> at revision:<revision>...
```

Reply:
> I think the west and manifest URL and revision should be determined by:

Just to clarify, this is all under the assumption that they're configurable via the west config file system.

There are other options as well, for example, we discussed using ``git config`` to set ``west.xxx`` variables in another PR.

Reply:
>There are other options as well, for example, we discussed using git config to set west.xxx variables in another PR.

I did not follow that discussion but it seems to me that if we are going to introduce a config file for west we might as well keep everything in it to avoid confusing users, who would then need to go search for options in both `~/.gitconfig` and `~/.westconfig`

Reply:
> I did not follow that discussion but it seems to me that if we are going to introduce a config file for west we might as well keep everything in it to avoid confusing users, who would then need to go search for options in both `~/.gitconfig` and `~/.westconfig`

I agree -- just pointing out that we did discuss alternative means to configure the remote URL and revision for these "special case" repositories in the past if it comes to it.

Reply:
This is not addressed yet by your patches right @ulfalizer?

Reply:
Is putting a manifest revision in a global configuration file just so that people can type `west init` instead of `west init -mr <rev>` really something we need in practice? It feels very obscure and a bit weird to me, like putting a global clone URL into `git` so that people can type `git clone` instead of `git clone <url>`.

Changing the manifest revision via `west init` would already be possible. You'd just rewrite the configuration file.

I'm imagining needing to rerun `west init` as something that would be pretty rare by the way, like rerunning `git init` (can do that to reinitialize hooks and stuff). Maybe you have a different workflow in mind.

Reply:
>Is putting a manifest revision in a global configuration file just so that people can type west init instead of west init -mr <rev> really something we need in practice?

Maybe I'm missing why we need a config file at all then? if we do have it, then I think we should allow people overriding the defaults. If we don't, then we should probably store these elsewhere?

Reply:
If we don't allow changing west's and the manifest's revision and url then we should do like git:

```
[remote "origin"]
        url = https://github.com/zephyrproject-rtos/west
        fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
        remote = origin
        merge = refs/heads/master
```

Those cannot be overridden by global config options I assume, so those are "local variables only"?

Reply:
So I would actually argue that we do it (again) just like Git does. Having some local-only options that are not read from the system or user config files because:

* I think Ulf has a point, I don't see why a revision or a URL should be permanently set in a configuration file, thinking about it again for the nth time. This feels like hardcoding the URL you pass to `git clone` or the branch that you pass to `git checkout`.
* Git does this by creating the config options above, we can do something similar so it is clear to the user that those are local-only
* We need to document the global options like Git does here: https://git-scm.com/docs/git-config so that the user knows what to expect

So in my view, I would approve this PR if:

1) We save both revision and URL in `west/config`
2) We document the Git options (maybe add a simple one like no colors)
3) We allow `west init` to be re-run after having changed the config file or passing new command-line params to it

Reply:
> * We save both revision and URL in `west/config`
> * We document the Git options (maybe add a simple one like no colors)
> * We allow `west init` to be re-run after having changed the config file or passing new command-line params to it

Sounds good to me.

Re-running `west init` seems alright to me, for the (I imagine, at least) rare cases where you'd need it. It'd just be a shorthand for changing the configuration.

Maybe we could add a `git config`-like at some point instead. I don't think it's horrible to have users directly edit configuration files in the meantime at least, regardless of the shorthand, as long as they're documented.

Reply:
> * I think Ulf has a point, I don't see why a revision or a URL should be permanently set in a configuration file, thinking about it again for the nth time. This feels like hardcoding the URL you pass to `git clone` or the branch that you pass to `git checkout`.

We do hardcode the git clone address of both west and the manifest, though, is the thing.

So there already is a default, and I don't see why the user should be prevented from doing a site- or user-wide override to `https://my-git-server.mycompany.com/{west,manifest}` or `https://github.com/myusername/{west,manifest}`.

> * Git does this by creating the config options above, we can do something similar so it is clear to the user that those are local-only

That said, I don't think this is worth blocking the merge over and I'm OK putting them into west/config if it's clear from the syntax that they are local to the wworkdir.

> We allow west init to be re-run after having changed the config file or passing new command-line params to it

Sounds like we're agreed then: re-running ``west init`` in an existing wworkdir (what I've been calling "reinit", not a separate command by that name) should be supported and allow changing the config so subsequent ``west fetch`` etc. reflect the new values.



> Maybe we could add a `git config`-like at some point instead. I don't think it's horrible to have users directly edit configuration files in the meantime at least, regardless of the shorthand, as long as they're documented.

I think editing config files by hand is perfectly natural FWIW. I actually prefer that for git config sometimes.